### PR TITLE
[DROOLS-6562] Manage cast for transformations/field refs

### PR DESCRIPTION
**Pull request by gitgabrio**
@danielezonca @kostola  @jiripetrlik

See https://issues.redhat.com/browse/DROOLS-6562

### CODE REVIEW

File: kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/expressions/KiePMMLConstant.java [hunk 1]
```diff
@@ -42,7 +45,8 @@ public Object getValue() {
 
     @Override
     public Object evaluate(final ProcessingDTO processingDTO) {
-        return getValue();
+        Object toReturn = getValue();
+        return dataType != null && toReturn != null ? dataType.getActualValue(toReturn) : toReturn;
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/expressions/KiePMMLConstant.java]: This code pattern is the same in many different place, what about move the logic inside a common method?

File: kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/KiePMMLOutputField.java [hunk 2]
```diff
@@ -93,8 +95,9 @@ public Object evaluate(final ProcessingDTO processingDTO) {
     }
 
     public Object evaluatePredictedValue(final ProcessingDTO processingDTO) {
-        return getValueFromKiePMMLNameValuesByVariableName(targetField, processingDTO.getKiePMMLNameValues())
+        Object toReturn = getValueFromKiePMMLNameValuesByVariableName(targetField, processingDTO.getKiePMMLNameValues())
                 .orElse(null);
+        return dataType != null && toReturn != null ? dataType.getActualValue(toReturn) : toReturn;
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/KiePMMLOutputField.java]: Here

File: kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/KiePMMLOutputField.java [hunk 3]
```diff
@@ -105,14 +108,15 @@ public Object evaluateReasonCodeValue(final ProcessingDTO processingDTO) {
             if (index < orderedReasonCodes.size()) {
                 resultCode = orderedReasonCodes.get(index);
             }
-           return resultCode;
+            return dataType != null && resultCode != null ? dataType.getActualValue(resultCode) : resultCode;
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/KiePMMLOutputField.java]: Here

File: kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/KiePMMLOutputField.java [hunk 4]
```diff
@@ -105,14 +108,15 @@ public Object evaluateReasonCodeValue(final ProcessingDTO processingDTO) {
...
         } else {
             return null;
         }
     }
 
     public Object evaluateTransformedValue(final ProcessingDTO processingDTO) {
-        return kiePMMLExpression != null ? kiePMMLExpression.evaluate(processingDTO) : null;
+        Object toReturn = kiePMMLExpression != null ? kiePMMLExpression.evaluate(processingDTO) : null;
+        return dataType != null && toReturn != null ? dataType.getActualValue(toReturn) : toReturn;
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/KiePMMLOutputField.java]: Here

File: kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/transformations/KiePMMLDefineFunction.java [hunk 5]
```diff
@@ -76,6 +78,7 @@ public Object evaluate(final ProcessingDTO processingDTO,
         for (KiePMMLNameValue kiePMMLNameValue : kiePMMLNameValues) {
             processingDTO.addKiePMMLNameValue(kiePMMLNameValue);
         }
-        return kiePMMLExpression.evaluate(processingDTO);
+        Object toReturn = kiePMMLExpression.evaluate(processingDTO);
+        return dataType != null && toReturn != null ? dataType.getActualValue(toReturn) : toReturn;
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/transformations/KiePMMLDefineFunction.java]: Here

File: kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/transformations/KiePMMLDerivedField.java [hunk 6]
```diff
@@ -69,15 +69,16 @@ public String getDisplayName() {
     }
 
     public Object evaluate(final ProcessingDTO processingDTO) {
-        return kiePMMLExpression.evaluate(processingDTO);
+        Object toReturn = kiePMMLExpression.evaluate(processingDTO);
+        return dataType != null && toReturn != null ? dataType.getActualValue(toReturn) : toReturn;
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/transformations/KiePMMLDerivedField.java]: Here

**Comment by danielezonca** [comment]
Jenkins retest this

**Comment by danielezonca** [comment]
Jenkins run cdb

File: kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/expressions/KiePMMLConstant.java (hunk 1 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/expressions/KiePMMLConstant.java]: @danielezonca 
Done

File: kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/KiePMMLOutputField.java (hunk 2 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/KiePMMLOutputField.java]: @danielezonca 
Done

File: kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/KiePMMLOutputField.java (hunk 3 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/KiePMMLOutputField.java]: @danielezonca 
Done

File: kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/KiePMMLOutputField.java (hunk 4 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/KiePMMLOutputField.java]: @danielezonca 
Done

File: kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/transformations/KiePMMLDefineFunction.java (hunk 5 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/transformations/KiePMMLDefineFunction.java]: @danielezonca 
Done

File: kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/transformations/KiePMMLDerivedField.java (hunk 6 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/transformations/KiePMMLDerivedField.java]: @danielezonca 
Done

**Review by danielezonca — APPROVED** [review]
LGTM 👍

**Review by jiripetrlik — APPROVED** [review]
LGTM

**Comment by gitgabrio** [comment]
@danielezonca 
Could this be merged ? The error I found in CDB seems unrelated

`java.lang.TypeNotPresentException: Type io.quarkus.maven.components.BootstrapSessionListener not present
	at org.eclipse.sisu.space.URLClassSpace.loadClass (URLClassSpace.java:147)
	at org.eclipse.sisu.space.NamedClass.load (NamedClass.java:46)
	at org.eclipse.sisu.space.AbstractDeferredClass.get (AbstractDeferredClass.java:48)
	...
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
 Caused by: java.lang.UnsupportedClassVersionError: io/quarkus/maven/components/BootstrapSessionListener has been compiled by a more recent version of the Java Runtime (class file version 55.0), this version of the Java Runtime only recognizes class file versions up to 52.0
	at java.lang.ClassLoader.defineClass1 (Native Method)
	at java.lang.ClassLoader.defineClass (ClassLoader.java:756)
	at java.security.SecureClassLoader.defineClass (SecureClassLoader.java:142)
	...`

**Pull request merged**